### PR TITLE
substitute solveWellEq with inner iterations and initial solve

### PR DESCRIPTION
**Pull request by totto82**
This clean-up has been on my list for a long time. It touches the initialization of the wells and would thus affect most cases. Hopefully to the better.

### CODE REVIEW

**Comment by totto82** [comment]
benchmark this please

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
jenkins build this please

**Comment by alfbr** [comment]
benchmark please

**Comment by ytelses** [comment]
Benchmark result overview: 




| Test    | Configuration                              | Relative |
| ------- | ------------------------------------------ | -------- |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1 | 0.569    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8 | 0.506    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1 | 1.044    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8 | 1.016    |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=1020

**Comment by totto82** [comment]
There seems to be some issues with the  "inner iterations" option for STW. It needs to be tested separately.

**Comment by totto82** [comment]
benchmark please

**Comment by ytelses** [comment]
Benchmark result overview: 




| Test    | Configuration                              | Relative |
| ------- | ------------------------------------------ | -------- |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1 | 0.956    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8 | 0.991    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1 | 0.93     |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8 | 0.93     |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=1022

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
benchmark please

**Comment by ytelses** [comment]
Benchmark result overview: 




| Test    | Configuration                              | Relative |
| ------- | ------------------------------------------ | -------- |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1 | 1.014    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8 | 1.024    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1 | 1.036    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8 | 0.976    |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=1026

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
benchmark please

**Comment by ytelses** [comment]
Benchmark result overview: 




| Test    | Configuration                              | Relative |
| ------- | ------------------------------------------ | -------- |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1 | 1.007    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8 | 1.016    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1 | 1.03     |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8 | 0.97     |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=1028

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
I have manually investigated the test failures. Failures are expected. The differences in results are small and acceptable IMO. Many models shows improvements in convergence (less iterations, problems etc.), but not all. See summary below. If we agree that this improves the logic of the code I think this is ready to get merged. 

```

    mpi.compareECLFiles_flow+SPE1CASE1 ok
    mpi.compareECLFiles_flow+SPE1CASE1_BRINE ok
    mpi.compareECLFiles_flow+SPE1CASE1_METRIC_VFP1 ok
    mpi.compareECLFiles_flow+3D_MSW ok
    mpi.compareECLFiles_flow+SPE9_CP_SHORT ok
    mpi.compareECLFiles_flow+SPE9_CP_GROUP ok
    mpi.compareECLFiles_flow+SPE5CASE1
    Keyword: IWEL, origin Restart, sequence 109

Value index                 = 318
(first value, second value) = (3, -1)

Non floating point values not identical

    mpi.compareECLFiles_flow+BASE_MODEL_1 ok
    mpi.compareECLFiles_flow+3D_WECON (59 != 58) ok
    mpi.compareECLFiles_flow+MSW_MODEL_1 ok 
    mpi.compareECLFiles_flow+0A1_GRCTRL_LRAT_ORAT_BASE_MODEL2_STW (10 != 9) ok
    mpi.compareECLFiles_flow+FAULTS_MODEL_1 ok
    mpi.compareECLFiles_flow+0A1_GRCTRL_LRAT_ORAT_BASE_MODEL2_MSW ok
    mpi.compareECLFiles_flow+0A2_GRCTRL_LRAT_ORAT_GGR_BASE_MODEL2_STW (11 != 10) ok
    mpi.compareECLFiles_flow+0A3_GRCTRL_LRAT_LRAT_BASE_MODEL2_STW (10 != 9) ok
    mpi.compareECLFiles_flow+0A2_GRCTRL_LRAT_ORAT_GGR_BASE_MODEL2_MSW ok
    mpi.compareECLFiles_flow+0A3_GRCTRL_LRAT_LRAT_BASE_MODEL2_MSW (23 != 22) ok
    mpi.compareECLFiles_flow+0A4_GRCTRL_LRAT_LRAT_GGR_BASE_MODEL2_STW (11 != 10) ok
    mpi.compareECLFiles_flow+0_BASE_MODEL2 (65 != 59) ok
    mpi.compareECLFiles_flow+0A4_GRCTRL_LRAT_LRAT_GGR_BASE_MODEL2_MSW (23 != 22) ok
    mpi.compareECLFiles_flow+UDQ_ACTIONX ok
    mpi.compareECLFiles_flow+UDQ_WCONPROD ok
    mpi.compareECLFiles_flow+UDQ_M3 ok
    mpi.compareECLFiles_flow+UDQ_M2 ok
    mpi.compareECLFiles_flow+PYACTION_WCONPROD  ok
    mpi.compareECLFiles_flow+UDQ_M1 (55 != 53) ok
    mpi.compareECLFiles_flow+1_MULTREGT_MODEL2 (66 != 67) ok
    mpi.compareECLFiles_flow+2_MULTXYZ_MODEL2 (58 != 55) ok
    mpi.compareECLFiles_flow+3_MULTFLT_MODEL2 (62 != 61) ok
    mpi.compareECLFiles_flow+4_MINPVV_MODEL2 (62 != 61) ok
    mpi.compareECLFiles_flow+5_SWATINIT_MODEL2 (63 != 65) ok
    mpi.compareECLFiles_flow+6_ENDSCALE_MODEL2 (62 != 64) ok
    mpi.compareECLFiles_flow+7_HYSTERESIS_MODEL2 (64 != 63) ok
    mpi.compareECLFiles_flow+9_1A_DEPL_MAX_RATE_MIN_BHP_STW (25 != 23) ok
    mpi.compareECLFiles_flow+8_MULTIPLY_TRANXYZ_MODEL2 (63 != 61) ok
    mpi.compareECLFiles_flow+9_EDITNNC_MODEL2 (64 != 65) ok 
    mpi.compareECLFiles_flow+9_1A_DEPL_MAX_RATE_MIN_BHP_MSW (29 != 23) ok
    mpi.compareECLFiles_flow+9_1B_DEPL_MAX_RATE_MIN_THP_STW (30 != 32) ok
    mpi.compareECLFiles_flow+9_2A_DEPL_GCONPROD_1L_STW (30 != 31) ok
    mpi.compareECLFiles_flow+9_2B_DEPL_GCONPROD_2L_STW (31 != 32) ok
    mpi.compareECLFiles_flow+9_2A_DEPL_GCONPROD_1L_MSW ok
    mpi.compareECLFiles_flow+9_3A_GINJ_REIN-G_STW ok
    mpi.compareECLFiles_flow+9_2B_DEPL_GCONPROD_2L_MSW ok
    mpi.compareECLFiles_flow+9_1B_DEPL_MAX_RATE_MIN_THP_MSW (31 != 32) ok
    mpi.compareECLFiles_flow+9_3B_GINJ_GAS_EXPORT_STW (32 != 34) ok
    mpi.compareECLFiles_flow+9_3A_GINJ_REIN-G_MSW ok
    mpi.compareECLFiles_flow+9_3C_GINJ_GAS_GCONSUMP_STW (32 != 31) ok
    mpi.compareECLFiles_flow+9_3B_GINJ_GAS_EXPORT_MSW ok
    mpi.compareECLFiles_flow+9_3D_GINJ_GAS_MAX_EXPORT_STW (32 != 35) ok
    mpi.compareECLFiles_flow+9_3E_GAS_MIN_EXPORT_STW  (25 != 24) ok
    mpi.compareECLFiles_flow+9_3C_GINJ_GAS_GCONSUMP_MSW (31 != 32) ok
    mpi.compareECLFiles_flow+9_3E_GAS_MIN_EXPORT_MSW ok(25 != 24) ok
    mpi.compareECLFiles_flow+9_4A_WINJ_MAXWRATES_MAXBHP_GCONPROD_1L_STW ok
    mpi.compareECLFiles_flow+9_3D_GINJ_GAS_MAX_EXPORT_MSW ok
    mpi.compareECLFiles_flow+9_4B_WINJ_VREP-W_STW (33 != 31) ok
    mpi.compareECLFiles_flow+9_4A_WINJ_MAXWRATES_MAXBHP_GCONPROD_1L_MSW ok-
    mpi.compareECLFiles_flow+9_4C_WINJ_GINJ_VREP-W_REIN-G_STW (41 != 40) ok
    mpi.compareECLFiles_flow+9_4B_WINJ_VREP-W_MSW ok 
    mpi.compareECLFiles_flow+9_4D_WINJ_GINJ_GAS_EXPORT_STW (42 != 40) ok
    mpi.compareECLFiles_flow+9_4C_WINJ_GINJ_VREP-W_REIN-G_MSW (32 != 33) ok
    mpi.compareECLFiles_flow+1_MSW_MODEL6 ok
    mpi.compareECLFiles_flow+9_4D_WINJ_GINJ_GAS_EXPORT_MSW (32 != 33) ok
    mpi.compareECLFiles_flow+TEST_WSEGSICD ok
    mpi.compareECLFiles_flow+BASE_MSW_WSEGAICD ok 
    mpi.compareECLFiles_flow+COMPL_SMRY ok
    mpi.compareECLFiles_flow+MOD4_UDQ_ACTIONX ok
    mpi.compareECLFiles.ok_flow+MOD4_GRP (116 != 105) ok
```

**Comment by totto82** [comment]
@GitPaean Can you review this.

**Comment by bska** [comment]
I've not assessed the other differences, but this diagnostic in particular

> ```
> mpi.compareECLFiles_flow+SPE5CASE1
> Keyword: IWEL, origin Restart, sequence 109
>
> Value index                 = 318
> (first value, second value) = (3, -1)
>
> Non floating point values not identical
>```

means that the third well ([`INJG`](https://github.com/OPM/opm-tests/blob/786cba4662aaac48c764b1301320a78edb1b763e/spe5/SPE5.BASE#L410)) switches from being active to being shut (strictly speaking, [no flowing connections](https://github.com/OPM/opm-common/blob/e185ddf96b0c16c79240a7b97a5e2d25d93d7f2c/src/opm/output/eclipse/AggregateWellData.cpp#L407-L408)) at the associate report step.  Are you sure that's okay?

**Comment by GitPaean** [comment]
> @GitPaean Can you review this.

I will try. But one of the the first things I will do is some testing running among different field models.  If you have tested different field models, also please report the findings here.

File: opm/simulators/wells/BlackoilWellModel_impl.hpp [hunk 1]
```diff
@@ -406,6 +406,18 @@ namespace Opm {
         //compute well guideRates
         const auto& comm = ebosSimulator_.vanguard().grid().comm();
         WellGroupHelpers::updateGuideRatesForWells(schedule(), phase_usage_, reportStepIdx, simulationTime, well_state_, comm, guideRate_.get());
+        updateAndCommunicateGroupData();
+        // Compute initial well solution for new wells
+        for (auto& well : well_container_) {
+            const uint64_t effective_events_mask = ScheduleEvents::WELL_STATUS_CHANGE
+                                                 + ScheduleEvents::NEW_WELL;
+            const bool event = report_step_starts_ && schedule().hasWellGroupEvent(well->name(), effective_events_mask, reportStepIdx);
```
Comment (GitPaean) [inline opm/simulators/wells/BlackoilWellModel_impl.hpp:414]: I got a compilation error related to `hasWellGroupEvent` in one computer. I could not find the definition of the function `hasWellGroupEvent`.  Not sure how it got compiled on Jenkins. Let me try again.

**Comment by GitPaean** [comment]
jenkins build this please

File: opm/simulators/wells/BlackoilWellModel_impl.hpp (hunk 1 above)
Comment (GitPaean) [reply opm/simulators/wells/BlackoilWellModel_impl.hpp:414]: It is confirmed in another setup. Please fix the compilation here.

Comment (bska) [reply opm/simulators/wells/BlackoilWellModel_impl.hpp:414]: > I got a compilation error related to `hasWellGroupEvent` in one computer.

The function was removed in @joakim-hove's recent PR OPM/opm-common#2194 ([diff](https://github.com/OPM/opm-common/pull/2194/files#diff-c4252a41c30bbb4d8fefb4e1c397779a3ec152c113be77ea238286d562bb8cb8L266)).  That PR in turn is part of a larger `Schedule` object refactoring (OPM/opm-common#2176).  This PR must be rebased onto the current master sources updated in the style of PR #3010.

Comment (totto82) [reply opm/simulators/wells/BlackoilWellModel_impl.hpp:414]: I guess it needs an rebase. See the recently merged
https://github.com/OPM/opm-simulators/pull/3010

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
@GitPaean I have rebased it so it should be ready for testing.

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
> I've not assessed the other differences, but this diagnostic in particular
> 
> mpi.compareECLFiles_flow+SPE5CASE1
> Keyword: IWEL, origin Restart, sequence 109
> 
> Value index                 = 318
> (first value, second value) = (3, -1)
> 
> Non floating point values not identical
> means that the third well (INJG) switches from being active to being shut (strictly speaking, no flowing connections) at the associate report step. Are you sure that's okay?

Yes. It is a WAG case where INJG, INJW opens/shuts in turn. The time-stepping are slightly different --> the index of when the well shuts/open is different. But if you look at the results in ResInsight they are spot on.

**Comment by totto82** [comment]
I think this both improves the code readably and the stability (since we now try to solve the well equation at the start of the well and not only at the start of the simulation). For many case this done not matter, but for MOD4_GRP and some of the NORNE prediction models this slightly improves the stability. It will need an update of the reference solution. If nobody object I will start the procedure of updating the reference solution later today.

**Comment by totto82** [comment]
jenkins build this update_data please

**Comment by bska** [comment]
jenkins build this opm-tests=472 please

**Comment by bska** [comment]
The updated reference solutions have been installed on the CI system.  I'll merge this now to bring the overall system back to a consistent state.  Thanks a lot for the contribution.

**Comment by bska** [comment]
@totto82: This morning I got a reproducible segmentation fault on a real field oil/water case that **may** be related to this PR.

In a debugger I traced the direct cause of the segmentation fault to indexing into an empty [`wellStateNupcol.wellRates()`](https://github.com/OPM/opm-simulators/blob/e77f8b2f7d3a9fbd4a7be0e34823691850814b6a/opm/simulators/wells/WellGroupHelpers.cpp#L337) array as part of [`updateAndCommunicateGroupData()`](https://github.com/totto82/opm-simulators/blob/36ba423d426849759553162382374d742f726e40/opm/simulators/wells/BlackoilWellModel_impl.hpp#L1349).  This, in turn, was triggered in the new call to [`updateAndCommunicateGroupData()`](https://github.com/totto82/opm-simulators/blob/36ba423d426849759553162382374d742f726e40/opm/simulators/wells/BlackoilWellModel_impl.hpp#L410) that is introduced as part of `beginTimeStep()` here.

As it happens the segmentation fault coincides with the first report/time step that opens a well, but the simulation has been running for 10 report steps and roughly 300 days at that point.  Moreover, the previous report step happens to have the following convergence history
```
Iter    MB(O)      MB(W)      CNV(O)     CNV(W) 
   0  1.004e-13  1.494e-10  1.421e-06  1.927e-04
   1  5.271e-12  4.653e-10  2.324e-08  1.391e-03
   [...]
  17  2.236e-12  4.639e-10  2.235e-08  2.475e-04
```
This is important because we end up in a case where the [`iterationIdx`](https://github.com/totto82/opm-simulators/blob/36ba423d426849759553162382374d742f726e40/opm/simulators/wells/BlackoilWellModel_impl.hpp#L1335) defined as
```C++
const int iterationIdx = ebosSimulator_.model().newtonMethod().numIterations();
```
gets the value `17`.  Furthermore, the case uses
```
NUPCOL
  5 /
```
meaning the [iteration index check](https://github.com/totto82/opm-simulators/blob/36ba423d426849759553162382374d742f726e40/opm/simulators/wells/BlackoilWellModel_impl.hpp#L1342-L1344) fails and we end up **not** updating/assigning [`well_state_nupcol_`](https://github.com/OPM/opm-simulators/blob/e77f8b2f7d3a9fbd4a7be0e34823691850814b6a/opm/simulators/wells/BlackoilWellModel.hpp#L314).  This, in turn, means that the "nupcol" well state remains in its zero-well configuration which explains indexing into the empty well rates array.

If I apply the change below, which is admittedly very hacky and I'd prefer a better solution, then I can make a sequential simulation run.  It's not quite complete yet, but it looks promising.  The parallel run (`mpirun -np 6`) stops quite early due to what appear to be singular well systems and zero potentials.  I have not tried to revert this PR.

---

```diff
diff --git a/opm/simulators/wells/BlackoilWellModel.hpp b/opm/simulators/wells/BlackoilWellModel.hpp
index 6790d3835..0ecb79b5e 100644
--- a/opm/simulators/wells/BlackoilWellModel.hpp
+++ b/opm/simulators/wells/BlackoilWellModel.hpp
@@ -386,7 +386,7 @@ namespace Opm {
 
             void updateWellControls(Opm::DeferredLogger& deferred_logger, const bool checkGroupControls);
 
-            void updateAndCommunicateGroupData();
+            void updateAndCommunicateGroupData(const bool beginTimeStep = false);
             void updateNetworkPressures();
 
             // setting the well_solutions_ based on well_state.
diff --git a/opm/simulators/wells/BlackoilWellModel_impl.hpp b/opm/simulators/wells/BlackoilWellModel_impl.hpp
index de650c09a..624595ad7 100644
--- a/opm/simulators/wells/BlackoilWellModel_impl.hpp
+++ b/opm/simulators/wells/BlackoilWellModel_impl.hpp
@@ -407,7 +407,7 @@ namespace Opm {
         const auto& comm = ebosSimulator_.vanguard().grid().comm();
         WellGroupHelpers::updateGuideRatesForWells(schedule(), phase_usage_, reportStepIdx, simulationTime, well_state_, comm, guideRate_.get());
         try {
-            updateAndCommunicateGroupData();
+            updateAndCommunicateGroupData(/* beginTimeStep = */ true);
             // Compute initial well solution for new wells
             for (auto& well : well_container_) {
                 const uint64_t effective_events_mask = ScheduleEvents::WELL_STATUS_CHANGE
@@ -1327,7 +1327,7 @@ namespace Opm {
     template<typename TypeTag>
     void
     BlackoilWellModel<TypeTag>::
-    updateAndCommunicateGroupData()
+    updateAndCommunicateGroupData(const bool beginTimeStep)
     {
         const int reportStepIdx = ebosSimulator_.episodeIndex();
         const Group& fieldGroup = schedule().getGroup("FIELD", reportStepIdx);
@@ -1339,7 +1339,7 @@ namespace Opm {
         const auto& comm = ebosSimulator_.vanguard().grid().comm();
         well_state_.updateGlobalIsGrup(schedule(), reportStepIdx, comm);
 
-        if (iterationIdx < nupcol) {
+        if (beginTimeStep || (iterationIdx < nupcol)) {
             well_state_nupcol_ = well_state_;
         }
```

**Comment by totto82** [comment]
@bska Thanks for a very thorough bug report. I am  a bit surprised that the iterationIdx is not reset when beginTimeStep is called. I will start digging immediately.

**Comment by totto82** [comment]
@bska Can you check if #3037 fix your problem. Yet another issue coming from gluing two simulators.
